xdsl/ir/core.py

```diff
@@ -573,38 +598,59 @@ def uses(self) -> IRUses:
 
     def add_use(self, use: Use):
         """Add a new use of the value."""
-        self._uses.add(use)
+        first_use = self.first_use
+        use._next_use = first_use  # pyright: ignore[reportPrivateUsage]
+        use._prev_use = None  # pyright: ignore[reportPrivateUsage]
+        if first_use is not None:
+            first_use._prev_use = use  # pyright: ignore[reportPrivateUsage]
+        self.first_use = use
 
     def remove_use(self, use: Use):
         """Remove a use of the value."""
-        assert use in self._uses, "use to be removed was not in use list"
-        self._uses.remove(use)
+        prev_use = use._prev_use  # pyright: ignore[reportPrivateUsage]
+        next_use = use._next_use  # pyright: ignore[reportPrivateUsage]
+        if prev_use is not None:
+            prev_use._next_use = next_use  # pyright: ignore[reportPrivateUsage]
+        if next_use is not None:
+            next_use._prev_use = prev_use  # pyright: ignore[reportPrivateUsage]
+
+        if prev_use is None:
+            self.first_use = next_use
 
     def has_one_use(self) -> bool:
         """Returns true if the value has exactly one use."""
-        return len(self._uses) == 1
+        first_use = self.first_use
+        return (
+            first_use is not None and first_use._next_use is None  # pyright: ignore[reportPrivateUsage]
+        )
 
     def has_more_than_one_use(self) -> bool:
         """Returns true if the value has more than one use."""
-        return len(self._uses) > 1
+        if (first_use := self.first_use) is None:
+            return False
+        return first_use._next_use is not None  # pyright: ignore[reportPrivateUsage]
 
     def get_unique_use(self) -> Use | None:
         """
         Returns the single use of the value, or None if there are no uses or
         more than one use.
         """
-        if self.has_one_use():
-            return next(iter(self._uses))
-        return None
+        if (first_use := self.first_use) is None:
+            return None
+        if first_use._next_use is not None:  # pyright: ignore[reportPrivateUsage]
+            return None
+        return first_use
 
     def get_user_of_unique_use(self) -> Operation | None:
         """
         Returns the user of the single use of the value.
         If there are no uses or more than one use, returns None.
         """
-        if (use := self.get_unique_use()) is not None:
-            return use.operation
-        return None
+        if (first_use := self.first_use) is None:
+            return None
+        if first_use._next_use is not None:  # pyright: ignore[reportPrivateUsage]
+            return None
+        return first_use.operation
```

Comment:
Why not stick with the above?

Reply:
Oh I think it happen when I changed it higher in the stack, let me put it back again

Reply:
Oh I think it happen when I changed it higher in the stack, let me put it back again